### PR TITLE
Implementing JsonSerializable for Teamspeak3_Helper_String

### PR DESCRIPTION
**Pull request by snegrini**
I'm currently developing a web app with this framework and I figured out it would be really useful to be able to return JSON encode of `Teamspeak3_Helper_String`.
Currently, if we use `json_encode()` on a `Teamspeak3_Helper_String` it will return an empty value `{ }`.
This implementation makes more comfortable, for example:
> getting info of clients and directly encode them with json_encode() (preparing a web page for AJAX requests).

EDIT: I wanted to make a pull request just for the last commit `d8e7b85`, but it seems I screwed things up and older commits are showing up.

### CODE REVIEW

**Comment by ronindesign** [comment]
Since this is a straight forward addition, would you be able to also include a unit test for this?
(I.e. something in `ts3phpframework/tests/Unit/Helper/StringTest.php`)

Thanks for this feature PR, I think it's a great idea.

**Comment by snegrini** [comment]
Yes, of course, I will add a test. In the meantime I was also trying to resolve the Failure of `StringTest::testIsInt()`, but I can't find a clue, yet.
About the JSON viewer, I have seen it, but it seemed to me it was far from what I really needed.
Though, I will give it a better look before pushing the test. Thank you for your time.

**Comment by svenpaulsen** [comment]
I like this. Thanks for contributing. @ronindesign Please merge this as soon as you think the unit test is fine.

**Comment by ronindesign** [comment]
Thanks for adding the test @snegrini, I appreciate it!

Regarding the `testIsInt()` failure, it's an outstanding problem resulting from a current issue in the code.
This issue is a bit meta and has to do with how the Server Query response comes in, so we're still working on a fix that is precise to the issue and doesn't over-assume anything on the part of the server's query protocol.

Anyways, thanks again, I'll review this and get it merged in this weekend.

File: libraries/TeamSpeak3/Helper/String.php [hunk 1]
```diff
@@ -850,6 +850,16 @@ public function __toString()
...
   }
 
+  /**
+   *  Specify data which should be serialized to JSON.
+   * 
+   * @return string
+   */
+  public function jsonSerialize()
+  {
+    return $this->string;
```
Comment (ronindesign) [inline libraries/TeamSpeak3/Helper/String.php]: Hmm, according to docs, it looks like string to encode must be unicode:

> value
> The value being encoded. Can be any type except a resource.
> **All string data must be UTF-8 encoded.**

[[PHP.net - json_encode](http://php.net/manual/en/function.json-encode.php)]

I think replacing with `return $this->toUtf8()` might be more appropriate.
Either that, or we need to force Unicode by throwing exception when string is not UTF-8 encoded.

Not sure if this is the right scope, but might be worth while to add a couple tests regarding expected behavior with/without a Unicode string. Maybe something like:

- 2nd test: check if UTF-8 string is still UTF-8 after encoding.
- 3rd test: check for exception when encoding fails, if possible.
(But these really belong in the scope of testing `toUtf8()` method, not here)

I'll add these in after I merge yours.

**Comment by snegrini** [comment]
I've currently made few more debug tests. It seems that json_encode() try to auto-call toUft8() method of the string. Here it is my debug code with an ASCII Valid 4 Octet Sequence:

```
$string = new \TeamSpeak3_Helper_String("\xf0\x90\x8c\xbc");

$a  = $string->toUtf8();
$b1 = json_encode(["test" => $string]);
$b2 = json_encode(["test" => $string], JSON_UNESCAPED_UNICODE);
$c1 = json_encode(["test" => $string->toUtf8()]);
$c2 = json_encode(["test" => $string->toUtf8()], JSON_UNESCAPED_UNICODE);
$d1 = json_encode(["test" => "\xf0\x90\x8c\xbc"]);
$d2 = json_encode(["test" => "\xf0\x90\x8c\xbc"], JSON_UNESCAPED_UNICODE);
```

 Variables content:

screen shot 2018-09-01 at 19 01 20

**Comment by ronindesign** [comment]
@snegrini Sorry for the delay in responding, I'm working on some tests for this, specifically want to make sure we're doing appropriate encoding as we don't want to lose / munge data unintentionally.

I've seen lots of reports of `utf8_encode()` being pretty abusive to non-standard data... so trying to take extra care to not introduce logic errors in the future.

**Comment by ronindesign** [comment]
@snegrini Hmm, maybe I misunderstand what you mean by:

> Here it is my debug code with an ASCII Valid 4 Octet Sequence

As far as I'm aware, the ASCII spec is defined as single, 8-bit byte, ranging over `x00` - `x7f` (i.e. 7-bit byte).
Thus, you example of `"\xf0\x90\x8c\xbc"` contains all characters outside the valid ASCII range.
Sorry if I'm confused here, encoding can be a bit of a tricky thing to discuss.

Here are some examples if this helps as well:

```
php > var_dump(mb_detect_encoding("\x00"));
string(5) "ASCII"

php > var_dump(mb_detect_encoding("\x5b"));
string(5) "ASCII"

php > var_dump(mb_detect_encoding("\x7f"));
string(5) "ASCII"

php > var_dump(mb_detect_encoding("\x80"));
bool(false)

php > var_dump(mb_detect_encoding("\xf0\x90\x8c\xbc"));
string(5) "UTF-8"

php > var_dump(mb_detect_encoding("\xf0\x90"));
string(5) "UTF-8"

php > var_dump(mb_detect_encoding("\xf0"));
string(5) "UTF-8"
```

**Comment by snegrini** [comment]
Nevermind, I was checking for Utf8 auto conversion of jsonSerialize().
I have been using this resource to submit the tests: https://www.cl.cam.ac.uk/~mgk25/ucs/examples/UTF-8-test.txt

**Comment by ronindesign** [comment]
Yeah, that list is great, I can definitely use that in building some good unit tests for boundary conditions.

So after doing some more testing, it looks like the `jsonSerialize()` is working correctly with the `toUtf8()` call, correct?

Let me know if you're still not seeing the correct encoding results on your end.

Otherwise, I'll finish up some unit tests and ship this PR! :shipit:

**Comment by snegrini** [comment]
Correct, I'll try some more but it seems working well.

**Comment by ronindesign** [comment]
Thanks for all the work on this @snegrini, sorry it took me a bit longer than expected to push this through!
